src/noise-models.lisp

```diff
@@ -0,0 +1,262 @@
+(in-package #:qvm)
+
+;;; A noise predicate is a function that takes an instruction's gate
+;;; and qubits, and returns either true if the instruction is a match,
+;;; else false. The priority indicates the priority of the predicate
+;;; (tbd) and the noise-position indicates whether the predicate
+;;; should match before or after an instruction.
+(defclass noise-pred ()
+  ((predicate
+    :initarg :predicate
+    :reader predicate
+    :initform nil
+    :documentation "Does the instruction trigger the application of a channel?")
+   (priority
+    :initarg :priority
+    :reader priority
+    :initform 10 ; Default priority is 10
+    :documentation "For ordering when checking for an instruction's matching predicates. The convention we use is that more specific predicates get higher priorities.")
+   (noise-position
+    :initarg :noise-position
+    :reader noise-position
+    :initform nil
+    :type noise-pos
+    :documentation "Should application of a channel happen before or after the instruction?")
+   (name
+    :initarg :name
+    :reader name
+    :type name
+    :documentation "Name or descriptor for noise."))
+  (:default-initargs :name +DEFAULT-NOISE-PRED-NAME+)
+  (:documentation "A NOISE-PRED describes a collection of program instructions. When used in a NOISE-RULE, the NOISE-PREDICATE describes which instructions should be matched by the noise described in the OPERATION-ELEMENTS. Optionally, the NOISE-PRED can be given a NAME. If none is provided, use the default 'NOISE-NAME'."))
+
+
+(defconstant +MAX-PRIORITY+ 20)
```

Comment:
move these constants to the top of the file imho

Reply:
also add comment